File: cinch/parser.py
import sys

from cinch_types import (Operator, If, While, IntegerLiteral, FunctionCall,
                         FunctionDef, Identifier, IdentifierList,
                         StatementList, ExpressionList, ArgumentList,
                         Operators)


# Now this is where things get complicated. The parser takes in an array of
# tokens and returns an abstract syntax tree. Parsing is perhaps the hardest
# problem in the process of compilation, but don't fear! We'll take it slowly
# and keep it simple.

def eat(expected_token, tokens):
    if tokens[0] != expected_token:
        print 'Syntax error'
        print 'eat', expected_token, tokens
        exit()
    tokens.pop(0)


# Parse takes a list of tokens and returns an abstract syntax tree.
def parse(tokens):
    return parse_statement_list(tokens)


# statement ::= expression | while_loop | if_statment
def parse_statement(tokens):
    # Gobble empty tokens
    if tokens[0] == '':
        tokens.pop(0)
        return parse_statement(tokens)
    elif tokens[0] == 'if':
        return parse_if_statement(tokens)
    elif tokens[0] == 'while':
        return parse_while_loop(tokens)
    else:
        return parse_single_expression(tokens)


# if_statement ::= 'if' '(' expression ')' '{' statement_list '}'
def parse_if_statement(tokens):
    tokens.pop(0)
    if_stmt = If('if')
    eat('(', tokens)
    condition = parse_expression(tokens)
    eat(')', tokens)
    eat('{', tokens)
    stmt_list = parse_statement_list(tokens)
    eat('}', tokens)
    if_stmt.children = [condition, stmt_list]
    return if_stmt


# while_loop ::= 'while' '(' expression ')' '{' statement_list '}'
def parse_while_loop(tokens):
    tokens.pop(0)
    while_loop = While('while')
    eat('(', tokens)
    condition = parse_expression(tokens)
    eat(')', tokens)
    eat('{', tokens)
    stmt_list = parse_statement_list(tokens)
    eat('}', tokens)
    while_loop.children = [condition, stmt_list]
    return while_loop


# integer_literal ::= [0-9]
def parse_integer_literal(tokens):
    if not tokens[0].isdigit():
        print("Syntax error")
        print 'int lit', tokens
        exit()
    i = int(tokens[0])
    tokens.pop(0)
    return IntegerLiteral(i)


# identifier ::= [a-zA-Z]
def parse_identifier(tokens):
    ident = tokens.pop(0)
    return Identifier(ident)


# identifier ::= [a-zA-Z]
def parse_identifier_list(tokens):
    id_list = IdentifierList()
    while tokens[0].isalpha():
        id_list.append(tokens.pop(0))
    return id_list


# operator ::= '=' | '+' | '-'
def parse_operator(tokens):
    op = tokens.pop(0)
    return Operator(op)


# function_call ::= identifier '(' argument_list ')'
def parse_function_call(tokens):
    func_call = FunctionCall(tokens[0])
    tokens.pop(0)
    eat('(', tokens)
    args = parse_argument_list(tokens)
    eat(')', tokens)
    func_call.children = [args]
    return func_call


def parse_function_def(tokens):
    """
    ```
    function_definition ::= 'function' identifier '(' identifier_list ')' '{'
                                statement_list
                            '}'
    ```
    """
    tokens.pop(0)
    func_def = FunctionDef(tokens[0])
    tokens.pop(0)
    eat('(', tokens)
    args = parse_argument_list(tokens)
    eat(')', tokens)
    eat('{', tokens)
    body = parse_statement_list(tokens)
    eat('}', tokens)
    func_def.children = [args, body]
    return func_def


def parse_expression(tokens):
    """
    expression ::= integer_literal
                 | binary_expression
                 | identifier
                 | function_call
    """
    expr = parse_single_expression(tokens)
    if expr is None:
        print 'Syntax error'
        print 'parse expr', tokens
        exit()
    return expr


def parse_single_expression(tokens):
    if len(tokens) >= 2 and tokens[1] == '(':
        expr = parse_function_call(tokens)
    elif tokens[0].isalpha():
        expr = parse_identifier(tokens)
    elif tokens[0].isdigit():
        expr = parse_integer_literal(tokens)
    else:
        return None
    if len(tokens) > 1 and tokens[0] in Operators:
        return parse_binary_expression(expr, tokens)
    else:
        return expr


def parse_expression_list(tokens):
    """Parse a list of expressions. Will return a ExpressionList object. The
    production for an expression list is:
    ```
    expression_list ::= expression expression_list | expression | epsilon
    ```"""
    expr_list = ExpressionList()
    while True:
        expr = parse_single_expression(tokens)
        if expr is None:
            break
        expr_list.append(expr)
    return expr_list


def parse_statement_list(tokens):
    """Parse a list of statements. Will return a StatementList object. The
    production for a statement list is:
    ```
    statement_list ::= statement statement_list | statement | epsilon
    ```"""
    stmt_list = StatementList()
    while True:
        stmt = parse_statement(tokens)
        if stmt is None:
            break
        stmt_list.append(stmt)
    return stmt_list


def parse_single_argument(tokens):
    """Parse a single argument. Will return None if the next token is not
    a valid argument.
    FIXME: all arguments list functions should be replaced with expression
    lists.
    """
    if len(tokens) > 1 and tokens[1] == '(':
        return parse_function_call(tokens)
    elif (tokens[0].isalpha() and tokens[0] != 'if' and
          tokens[0] != 'while' and tokens[0] != 'function'):
        return parse_identifier(tokens)
    elif tokens[0].isdigit():
        return parse_integer_literal(tokens)
    else:
        return None


def parse_argument(tokens):
    """Parse a single argument. May raise SystemExit if the next token is not
    a valid argument.
    The production for an argument is:
    ```
    argument ::= identifier | integer_literal | function_call
    ```
    """
    arg = parse_single_argument(tokens)
    if arg is None:
        print 'Syntax error'
        print 'parse_args', tokens
        sys.exit()
    return arg


def parse_argument_list(tokens):
    """Parse a list of arguments. `tokens` is the list of tokens.
    A list of arguments is zero or more arguments. An argument is an identifier
    or an integer literal.
    The production for a list of arguments is:
    ```
    argument_list ::= argument
                    | argument argument_list
                    | epsilon
    ```
    """
    arg_list = ArgumentList()
    while True:
        arg = parse_single_argument(tokens)
        if arg is None:
            break
        arg_list.append(arg)
    return arg_list


def parse_binary_expression(lhs, tokens):
    """Parse a binary expression. `lhs` is an expression representing the left
    hand side of the binary expression. `tokens` is a list of tokens.
    May raise SystemExit.
    The production for binary expressions is:
    ```
    binary_expression ::= integer_literal operator expression
                        | identifier operator expression
                        | function_call operator expression
    ```
    """
    operator = Operator(tokens.pop(0), [lhs])
    rhs = parse_expression(tokens)
    operator.children.append(rhs)
    return operator

File: cinchs/Parser.hs
module Parser where

import Data.Char
import CinchTypes

-- TODO: Write Parser monad

parse :: [String] -> Maybe Ast
parse tokens = case parseStatementList tokens of
  Just (ast, _) -> Just ast
  Nothing -> Nothing

parseStatementList :: [String] -> Maybe (Ast, [String])
parseStatementList tokens = case parseStatementListHelper [] tokens of
  Nothing -> Nothing
  Just (sl, others) -> Just (Ast { typ=StatementList, name="", children=sl }, others)

parseStatementListHelper :: [Ast] -> [String] -> Maybe ([Ast], [String])
parseStatementListHelper asts [] = Just (reverse asts, [])
parseStatementListHelper asts ("}":tokens) = Just (reverse asts, tokens)
parseStatementListHelper asts tokens = case parseStatement tokens of
  Nothing -> Nothing
  Just (ast, rest) -> parseStatementListHelper (ast:asts) rest



parseExpressionList :: [String] -> Maybe (Ast, [String])
parseExpressionList tokens = case parseExpressionListHelper [] tokens of
  Nothing -> Nothing
  Just (sl, others) -> Just (Ast { typ=ExpressionList, name="", children=sl }, others)

parseExpressionListHelper :: [Ast] -> [String] -> Maybe ([Ast], [String])
parseExpressionListHelper asts [] = Just (reverse asts, [])
parseExpressionListHelper asts (")":tokens) = Just (reverse asts, tokens)
parseExpressionListHelper asts tokens = case parseExpression tokens of
  Nothing -> Nothing
  Just (ast, rest) -> parseExpressionListHelper (ast:asts) rest



parseStatement :: [String] -> Maybe (Ast, [String])
parseStatement ("if":"(":tokens) = parseIfStatement tokens
parseStatement ("while":"(":tokens) = parseWhileLoop tokens
parseStatement ("function":tokens) = parseFuncDef tokens
parseStatement ("return":tokens) = parseReturnStatement tokens
parseStatement (variable:"=":tokens) = parseAssignment (variable:"=":tokens)
parseStatement tokens = parseExpression tokens

parseAssignment :: [String]  -> Maybe (Ast, [String])
parseAssignment tokens = case parseIdentifier tokens of
  Nothing -> Nothing
  Just (name, "=":rest) -> case parseExpression rest of
    Nothing -> Nothing
    Just (expr, others) -> Just (Ast { typ=AssignmentStatement, name="", children=[name, expr] }, others)
  otherwise -> Nothing

parseIfStatement :: [String] -> Maybe (Ast, [String])
parseIfStatement tks = case parseExpression tks of
  Nothing -> Nothing
  Just (cond, ")":body) -> case parseBlock body of
    Nothing -> Nothing
    Just (body, rest) -> Just (Ast { typ=If, name="", children=[cond, body] }, rest)

parseWhileLoop :: [String] -> Maybe (Ast, [String])
parseWhileLoop tks = case parseExpression tks of
  Nothing -> Nothing
  Just (cond, ")":body) -> case parseBlock body of
    Nothing -> Nothing
    Just (body, rest) -> Just (Ast { typ=While, name="", children=[cond, body] }, rest)

parseFuncDef :: [String] -> Maybe (Ast, [String])
parseFuncDef (name:"(":tokens) = case parseIdentifierList tokens of
  Nothing -> Nothing
  Just (argumentNames, newtoks) -> case parseBlock newtoks of
    Nothing -> Nothing
    Just (body, newertoks) -> Just (Ast { typ=FunctionDef, name=name, children=argumentNames++[body] }, newertoks)
parseFuncDef otherwise = Nothing

parseReturnStatement :: [String] -> Maybe (Ast, [String])
parseReturnStatement rest = case parseExpression rest of
  Nothing -> Nothing
  Just (expr, others) -> Just (Ast { typ=Return, name="", children=[expr] }, others)

parseExpression :: [String] -> Maybe (Ast, [String])
parseExpression tokens = case parseExpressionHelper tokens of
  Nothing -> Nothing
  Just (first, []) -> Just (first, [])
  Just (first, op:rest) -> if op `elem` operators then
      parseBinaryExpression first (op:rest)
    else
      Just (first, op:rest)

parseExpressionHelper :: [String] -> Maybe (Ast, [String])
parseExpressionHelper (f:"(":tokens) = parseFuncCall (f:"(":tokens)
parseExpressionHelper tokens
  | all isDigit $ head tokens = parseNumber tokens
  | all isAlpha $ head tokens = parseIdentifier tokens
  | otherwise = Nothing

parseIdentifierList :: [String] -> Maybe ([Ast], [String])
parseIdentifierList = parseIdentifierListHelper []

parseIdentifierListHelper :: [Ast] -> [String] -> Maybe ([Ast], [String])
parseIdentifierListHelper asts (")":rest) = Just (reverse asts, rest)
parseIdentifierListHelper asts tokens = case parseIdentifier tokens of
  Nothing -> Nothing
  Just (ast, rest) -> parseIdentifierListHelper (ast:asts) rest

parseFuncCall :: [String] ->  Maybe (Ast, [String])
parseFuncCall tokens = case parseIdentifier tokens of
  Nothing -> Nothing
  Just (name, "(":rest) -> case parseExpressionList rest of
    Nothing -> Nothing
    Just (args, others) -> Just (
      Ast { typ=FunctionCall, name="", children=[name, args] },
      others)
  otherwise -> Nothing

parseNumber :: [String] ->  Maybe (Ast, [String])
parseNumber (name:tokens) = Just (Ast { typ=IntegerLiteral, name=name, children=[] }, tokens)

parseIdentifier :: [String] -> Maybe (Ast, [String])
parseIdentifier (name:tokens) = Just (Ast { typ=Identifier, name=name, children=[] }, tokens)

parseBinaryExpression :: Ast -> [String] -> Maybe (Ast, [String])
parseBinaryExpression left (operator:tokens) = case parseExpression tokens of
  Nothing -> Nothing
  Just (right, rest) -> Just (Ast { typ=BinaryExpression, name=operator, children=[left, right] }, rest)

parseBlock :: [String] -> Maybe (Ast, [String])
parseBlock ("{":body) = case parseStatementList body of
  Nothing -> Nothing
  Just (ast, rest) -> Just (ast, rest)
parseBlock otherwise = Nothing
